python/taichi/lang/source_builder.py

```diff
@@ -0,0 +1,138 @@
+import atexit
+import ctypes
+import os
+import shutil
+import subprocess
+import tempfile
+
+from taichi.core.util import ti_core as _ti_core
+from taichi.lang import impl
+from taichi.lang.exception import TaichiSyntaxError
+from taichi.lang.expr import make_expr_group
+from taichi.lang.util import get_clangpp, has_clangpp
+
+
+class SourceBuilder:
+    def __init__(self):
+        self.bc = None
+        self.so = None
+        self.mode = None
+        self.td = None
+
+        def cleanup():
+            if self.td is not None:
+                shutil.rmtree(self.td)
+
+        atexit.register(cleanup)
+
+    @classmethod
+    def from_file(cls, filename, compile_fn=None, _temp_dir=None):
+        self = cls()
+        self.td = _temp_dir
+        if filename.endswith((".cpp", ".c", ".cc")):
+            assert impl.current_cfg().arch in [
+                _ti_core.Arch.x64, _ti_core.Arch.cuda
+            ]
+            if compile_fn is None:
+                if self.td is None:
+                    self.td = tempfile.mkdtemp()
+
+                def compile_fn_impl(filename):
+                    if impl.current_cfg().arch == _ti_core.Arch.x64:
+                        subprocess.call(get_clangpp() + ' -flto -c ' +
+                                        filename + ' -o ' +
+                                        os.path.join(self.td, 'source.bc'),
+                                        shell=True)
+                    else:
+                        subprocess.call(get_clangpp() + ' -flto -c ' +
```

Comment:
What is the intent to compile a regular `cpp` file to `nvptx64`? Will they be marked as device functions by default?

Reply:
Yes. With ".cpp", all functions will be marked as device functions.
However, with ".cu", the user has to manually mark device function and other kinds.